### PR TITLE
Types unification

### LINKED ISSUE #217: Unify types.
- Merge `odra-casper-types` into `odra-types`.
- Use `odra-types` in `mock-vm`.
- Use `bytesrepr::{ToBytes, FromBytes}` instead of `OdraType`.
- Remove `Balance`. `attached_value` should be `U512` in both `casper` and `mock-vm`.

### PR DESCRIPTION
**Pull request by kpob**
Resolves #217

### CODE REVIEW

File: core/src/mapping.rs [hunk 1]
```diff
@@ -17,7 +22,7 @@ pub struct Mapping<K, V> {
     namespace_buffer: Vec<u8>
 }
 
-impl<K: OdraType, V: OdraType> Mapping<K, V> {
+impl<K: ToBytes, V: ToBytes + FromBytes + CLTyped> Mapping<K, V> {
```
Comment (zie1ony) [inline core/src/mapping.rs]: Looks like `V` could be just `FromBytes + CLTyped`.

File: core/src/test_utils.rs [hunk 2]
```diff
@@ -1,13 +1,17 @@
...
+    event::OdraEvent,
+    Address
+};
 
 /// Gets the nth event emitted by the contract at `address`.
 ///
 /// If the passed index is out of bounds, or a deserialization error occurs,
 /// an error is returned.
-pub fn get_event<T: OdraType + OdraEvent>(
+pub fn get_event<T: ToBytes + FromBytes + OdraEvent>(
```
Comment (zie1ony) [inline core/src/test_utils.rs]: Value that is being loaded doesn't have to be `ToBytes`.

File: examples/src/contracts/tlw.rs [hunk 3]
```diff
@@ -1,12 +1,12 @@
 use odra::{
     contract_env, execution_error,
-    types::{event::OdraEvent, Address, Balance, BlockTime},
+    types::{casper_types::U512, event::OdraEvent, Address, BlockTime},
```
Comment (zie1ony) [inline examples/src/contracts/tlw.rs:3]: I think reexporting `U256` and `U512` in `odra-types` would simplify imports.

File: mock-vm/backend/src/test_env.rs [hunk 4]
```diff
@@ -67,33 +68,34 @@ where
...
-    index: i32
-) -> Result<T, EventError> {
+pub fn get_event<T: FromBytes + OdraEvent>(address: Address, index: i32) -> Result<T, EventError> {
     let bytes = crate::borrow_env().get_event(address, index);
 
     bytes.and_then(|bytes| {
         let event_name = extract_event_name(&bytes)?;
-        if event_name == T::name() {
-            T::deser(bytes).map_err(|_| EventError::Parsing)
+        dbg!(&event_name);
```
Comment (zie1ony) [inline mock-vm/backend/src/test_env.rs]: dbg

File: odra-casper/codegen/src/ty/odra.rs [hunk 5]
```diff
@@ -1,70 +1,67 @@
-use odra_types::Type;
+use odra_types::casper_types::CLType;
 use proc_macro2::TokenStream;
 use quote::{quote, ToTokens};
 
-pub struct OdraType<'a>(&'a Type);
+pub struct OdraType<'a>(&'a CLType);
```
Comment (zie1ony) [inline odra-casper/codegen/src/ty/odra.rs:5]: Do we need `OdraType` anymore?

File: types/Cargo.toml [hunk 6]
```diff
@@ -4,11 +4,18 @@ edition = "2021"
...
+
+[dependencies]
+casper-types = { workspace = true }
 
 [dev-dependencies]
 odra-utils = { path = "../utils", version = "0.6.0" }
+
+[features]
+default = []
+test-support = []
```
Comment (zie1ony) [inline types/Cargo.toml:21]: What is `test-support`?

File: types/src/address.rs [hunk 7]
```diff
@@ -1,3 +1,290 @@
+//! Better address representation for Casper.
+
+use core::str::FromStr;
+
+use crate::AddressError;
+use crate::AddressError::ZeroAddress;
+use crate::{OdraError, VmError};
+use alloc::{
```
Comment (zie1ony) [inline types/src/address.rs:8]: use `crate::prelude`

File: types/src/error.rs [hunk 8]
```diff
@@ -50,6 +50,20 @@ impl From<Box<dyn Any + Send>> for OdraError {
     }
 }
 
+impl From<casper_types::bytesrepr::Error> for ExecutionError {
```
Comment (zie1ony) [inline types/src/error.rs:54]: Do we need to discuss that?

File: types/src/uints.rs [hunk 9]
```diff
@@ -0,0 +1,38 @@
+use casper_types::{U256, U512};
+
+use crate::arithmetic::ArithmeticsError;
+
+pub trait ToU256 {
```
Comment (zie1ony) [inline types/src/uints.rs:5]: +1

**Review by zie1ony — APPROVED** [review]

File: types/src/address.rs (hunk 7 above)
Comment (kpob) [reply types/src/address.rs:8]: prelude is implemented in `core` so all the other crates explicitly use `alloc`.

File: odra-casper/codegen/src/ty/odra.rs (hunk 5 above)
Comment (kpob) [reply odra-casper/codegen/src/ty/odra.rs:5]: It's a part of casper codegen, not the trait we used to use.

File: examples/src/contracts/tlw.rs (hunk 3 above)
Comment (kpob) [reply examples/src/contracts/tlw.rs:3]: Makes sense

File: types/Cargo.toml (hunk 6 above)
Comment (kpob) [reply types/Cargo.toml:21]: Cause both casper and mock-vm use the same, casper implementation of Address, it's handy to have some utility functions to create an Address (both account and contract).
The former implementation of Address for mock-vm allowed to create an address like `Address::new("alice")`. So `test-support` feature provides function to create addresses in test env.

File: types/src/error.rs (hunk 8 above)
Comment (kpob) [reply types/src/error.rs:54]: I would implement it like
```rust
impl From<casper_types::bytesrepr::Error> for ExecutionError {
    fn from(value: casper_types::bytesrepr::Error) -> Self {
        Self::sys(CODE_SERIALIZATION_FAILED, match value {
            casper_types::bytesrepr::Error::EarlyEndOfStream => "Early end of stream",
            casper_types::bytesrepr::Error::Formatting => "Formatting",
            casper_types::bytesrepr::Error::LeftOverBytes => "Leftover bytes",
            casper_types::bytesrepr::Error::OutOfMemory => "Out of memory",
            casper_types::bytesrepr::Error::NotRepresentable => "Not representable",
            casper_types::bytesrepr::Error::ExceededRecursionDepth => "Exceeded recursion depth",
            _ => "Serialization failed"
        })
    }
}
``

File: types/Cargo.toml (hunk 6 above)
Comment (zie1ony) [reply types/Cargo.toml:21]: We could leave it this way, but later I see this more in `cfg(test)` context rather via feature.

Comment (kpob) [reply types/Cargo.toml:21]: But if apply the config flag, its scope would be the current crate only, so if it is implemented  in `types`, you couldn't  use it somewhere else (in `mockvm` in this case)

File: types/src/address.rs (hunk 7 above)
Comment (zie1ony) [reply types/src/address.rs:8]: You're right.

File: types/src/error.rs (hunk 8 above)
Comment (zie1ony) [reply types/src/error.rs:54]: +1

**Pull request merged**